File: PetAgent/Avatar/AvatarManifest.swift
//
//  AvatarManifest.swift
//  PetAgent
//
//  F2 · owner: 강상우 (Sangwoo Kang)
//  manifest.json Codable model (mirrors protocol repo section 6 schema).
//

/// A value in the clips table.
///
/// For `type: usdz`, `.name(String)` is the file stem of a **single-animation**
/// usdz living alongside manifest.json (e.g. "idle" -> Avatars/{name}/idle.usdz)
/// — NOT an animation name to look up inside one shared model. RealityKit
/// effectively only plays a usdz's first animation regardless of how many
/// `availableAnimations` entries it reports, so one avatar with 10 named clips
/// baked into a single usdz does not work; each clip needs its own file. See
/// pet-app/docs/avatar-spec.md for the full external-creator requirements.
///
/// `type: sprites` reuses the same file-stem convention. `type: video` uses
/// a {"in":sec,"out":sec} time range instead (`.timeRange`).
enum ClipReference: Equatable {
    case name(String)
    case timeRange(in: Double, out: Double)
}

extension ClipReference: Codable {
    private enum CodingKeys: String, CodingKey {
        case `in`, out
    }

    init(from decoder: Decoder) throws {
        if let container = try? decoder.container(keyedBy: CodingKeys.self),
           let start = try? container.decode(Double.self, forKey: .in),
           let end = try? container.decode(Double.self, forKey: .out) {
            self = .timeRange(in: start, out: end)
            return
        }
        let single = try decoder.singleValueContainer()
        self = .name(try single.decode(String.self))
    }

    func encode(to encoder: Encoder) throws {
        switch self {
        case .name(let name):
            var container = encoder.singleValueContainer()
            try container.encode(name)
        case .timeRange(let start, let end):
            var container = encoder.container(keyedBy: CodingKeys.self)
            try container.encode(start, forKey: .in)
            try container.encode(end, forKey: .out)
        }
    }
}

struct AvatarManifest: Equatable, Codable {
    enum AvatarType: String, Codable {
        case usdz
        case video
        case sprites
    }

    struct Hitbox: Equatable, Codable {
        let width: Double
        let height: Double
    }

    let schemaVersion: Int
    let name: String
    let type: AvatarType
    let scale: Double
    let hitbox: Hitbox
    let clips: [String: ClipReference]
    let sounds: [String: String]

    private enum CodingKeys: String, CodingKey {
        case schemaVersion = "schema_version"
        case name, type, scale, hitbox, clips, sounds
    }
}

File: src/types/avatar-manifest.ts
/**
 * Avatar manifest.json schema (protocol docs/avatar-manifest.md, plan/01_protocol.md
 * section 6). Mirrors pet-app/PetAgent/Avatar/AvatarManifest.swift.
 */

export type AvatarType = "usdz" | "video" | "sprites";

export interface Hitbox {
  width: number;
  height: number;
}

/**
 * A value in the clips table.
 *
 * For `type: usdz`/`sprites`, a plain string is a file stem living alongside
 * manifest.json (e.g. "walk" -> Avatars/{name}/walk.usdz) -- NOT an animation
 * name inside one shared model. RealityKit effectively only plays a usdz's
 * first animation regardless of how many are baked in, so this is one usdz
 * file per clip, not one usdz with many named animations. See
 * pet-app/docs/avatar-spec.md for the full external-creator requirements.
 *
 * For `type: video`, the value is instead a { in, out } timecode range (seconds).
 */
export type ClipReference = string | { in: number; out: number };

/** Missing clips other than these fall back to idle; these two are mandatory. */
export const REQUIRED_CLIPS = ["idle", "walk"] as const;

/**
 * sounds keys mix clip names (played on FSM state entry) and event names
 * (played on socket events, protocol events.ts BridgeEvent). Both lookups
 * hit this same table. The event-name keys pair with the events those names
 * name -- e.g. "await_approval" is the SFX for the await_approval BridgeEvent.
 */
export type SoundEventKey =
  | "app_launch"
  | "task_success"
  | "task_fail"
  | "await_approval"
  | "listen_start";

export interface AvatarManifest {
  schema_version: number;
  name: string;
  /** First implementation only supports "usdz"; "video"/"sprites" are reserved. */
  type: AvatarType;
  /**
   * Multiplier applied to the root entity so that raw mesh height * scale ~= 1
   * unit (1 unit == 100 screen px). Prefer normalizing the source mesh to 1
   * unit (scale = 1.0) -- scale here is a correction fallback, not the primary
   * mechanism (scale = 1 / raw mesh height).
   */
  scale: number;
  hitbox: Hitbox;
  clips: Record<string, ClipReference>;
  /** Keys are a mix of clip names and SoundEventKey values -- see SoundEventKey. */
  sounds: Record<string, string>;
}
